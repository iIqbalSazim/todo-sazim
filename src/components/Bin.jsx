import { ActionIcon, Group, Menu } from "@mantine/core";
import { IconTrashFilled } from "@tabler/icons-react";

const Bin = ({ trash, handleRetrieveAllClick, handleEmptyBinClick }) => {
  return (
    <Group position="right" my={"xl"}>
      {trash.length !== 0 ? (
        <Menu width={200} shadow="md" withArrow position="left" offset={0}>
          <Menu.Target>
            <ActionIcon size={"4rem"} color="red.4" variant="transparent">
              <IconTrashFilled stroke="1px" size={"3rem"} />
            </ActionIcon>
          </Menu.Target>

          <Menu.Dropdown>
            <Menu.Item onClick={() => handleRetrieveAllClick()}>
              Retrieve all
            </Menu.Item>

            <Menu.Item onClick={() => handleEmptyBinClick()}>
              Empty Bin
            </Menu.Item>
          </Menu.Dropdown>
        </Menu>
      ) : null}
    </Group>
  );
};

export default Bin;
