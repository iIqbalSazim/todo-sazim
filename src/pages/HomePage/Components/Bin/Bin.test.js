import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Bin from "./Bin";
import userEvent from "@testing-library/user-event";

const mockTrash = [
  {
    id: 1,
    description: "Test task 1",
  },
  {
    id: 2,
    description: "Test task 2",
  },
];
const mockRetrieveAllClick = jest.fn();
const mockEmptyBinClick = jest.fn();

describe("Bin", () => {
  test("renders correctly", () => {
    const { container } = render(
      <Bin
        trash={mockTrash}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");

    expect(binButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("menu renders correctly after clicking on bin button", async () => {
    const { container } = render(
      <Bin
        trash={mockTrash}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");

    await userEvent.click(binButton);

    const retrieveAllMenuButton = screen.getByText(/retrieve all/i);
    const emptyBinMenuButton = screen.getByText(/empty bin/i);

    expect(retrieveAllMenuButton).toBeInTheDocument();
    expect(emptyBinMenuButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test("calls handleRetrieveAllClick on retrieve all button click", async () => {
    render(
      <Bin
        trash={mockTrash}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");
    await userEvent.click(binButton);

    const retrieveAllMenuButton = screen.getByText(/retrieve all/i);
    await userEvent.click(retrieveAllMenuButton);

    expect(mockRetrieveAllClick).toHaveBeenCalled();
  });

  test("calls handleEmptyBinClick on empty bin button click", async () => {
    render(
      <Bin
        trash={mockTrash}
        handleRetrieveAllClick={mockRetrieveAllClick}
        handleEmptyBinClick={mockEmptyBinClick}
      />
    );

    const binButton = screen.getByRole("button");
    await userEvent.click(binButton);

    const emptyBinMenuButton = screen.getByText(/empty bin/i);
    await userEvent.click(emptyBinMenuButton);

    expect(mockEmptyBinClick).toHaveBeenCalled();
  });
});
