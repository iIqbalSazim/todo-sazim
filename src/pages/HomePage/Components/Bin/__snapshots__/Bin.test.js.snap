// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bin menu renders correctly after clicking on bin button 1`] = `
<div>
  <div
    class="mantine-Group-root mantine-qokdfp"
  >
    <button
      aria-controls="mantine-hg8zm9qiy-dropdown"
      aria-expanded="true"
      aria-haspopup="menu"
      class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-y99nq2"
      data-expanded="true"
      id="mantine-hg8zm9qiy-target"
      type="button"
    >
      <svg
        class="tabler-icon tabler-icon-trash-filled"
        fill="none"
        height="2rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1px"
        viewBox="0 0 24 24"
        width="2rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6a1 1 0 0 1 .117 1.993l-.117 .007h-.081l-.919 11a3 3 0 0 1 -2.824 2.995l-.176 .005h-8c-1.598 0 -2.904 -1.249 -2.992 -2.75l-.005 -.167l-.923 -11.083h-.08a1 1 0 0 1 -.117 -1.993l.117 -.007h16z"
          fill="currentColor"
          stroke-width="0"
        />
        <path
          d="M14 2a2 2 0 0 1 2 2a1 1 0 0 1 -1.993 .117l-.007 -.117h-4l-.007 .117a1 1 0 0 1 -1.993 -.117a2 2 0 0 1 1.85 -1.995l.15 -.005h4z"
          fill="currentColor"
          stroke-width="0"
        />
      </svg>
    </button>
    <div
      aria-labelledby="mantine-hg8zm9qiy-target"
      aria-orientation="vertical"
      class="mantine-Menu-dropdown mantine-1nbod7e"
      data-position="left"
      id="mantine-hg8zm9qiy-dropdown"
      role="menu"
      style="transition-property: opacity; transition-duration: 150ms; transition-timing-function: ease; opacity: 1; z-index: 300; top: 0px; left: -3.5px; width: 12.5rem;"
      tabindex="-1"
    >
      <div
        data-autofocus="true"
        data-menu-dropdown="true"
        style="outline: 0;"
        tabindex="-1"
      >
        <button
          class="mantine-Menu-item mantine-eby8v3"
          data-menu-item="true"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <div
            class="mantine-1rnkhfs mantine-Menu-itemLabel"
          >
            Retrieve all
          </div>
        </button>
        <button
          class="mantine-Menu-item mantine-eby8v3"
          data-menu-item="true"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <div
            class="mantine-1rnkhfs mantine-Menu-itemLabel"
          >
            Empty Bin
          </div>
        </button>
      </div>
      <div
        class="mantine-ndinzw mantine-Menu-arrow"
        style="width: 0.4375rem; height: 0.4375rem; transform: rotate(45deg); position: absolute; border-top-right-radius: 0rem; right: -0.21875rem; border-left-color: transparent; border-bottom-color: transparent; top: -0.21875px;"
      />
    </div>
  </div>
</div>
`;

exports[`Bin renders correctly 1`] = `
<div>
  <div
    class="mantine-Group-root mantine-qokdfp"
  >
    <button
      aria-controls="mantine-n930dvqtn-dropdown"
      aria-expanded="false"
      aria-haspopup="menu"
      class="mantine-UnstyledButton-root mantine-ActionIcon-root mantine-y99nq2"
      id="mantine-n930dvqtn-target"
      type="button"
    >
      <svg
        class="tabler-icon tabler-icon-trash-filled"
        fill="none"
        height="2rem"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1px"
        viewBox="0 0 24 24"
        width="2rem"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20 6a1 1 0 0 1 .117 1.993l-.117 .007h-.081l-.919 11a3 3 0 0 1 -2.824 2.995l-.176 .005h-8c-1.598 0 -2.904 -1.249 -2.992 -2.75l-.005 -.167l-.923 -11.083h-.08a1 1 0 0 1 -.117 -1.993l.117 -.007h16z"
          fill="currentColor"
          stroke-width="0"
        />
        <path
          d="M14 2a2 2 0 0 1 2 2a1 1 0 0 1 -1.993 .117l-.007 -.117h-4l-.007 .117a1 1 0 0 1 -1.993 -.117a2 2 0 0 1 1.85 -1.995l.15 -.005h4z"
          fill="currentColor"
          stroke-width="0"
        />
      </svg>
    </button>
  </div>
</div>
`;
